---
title: "Differential expression downstream analysis shiny app"
package: dexdash
output: rmarkdown::html_vignette
abstract: |
  This R package, designed for interactive exploration and analysis
    of differential expression or abundance data, integrates Shiny-based
    visualization tools with rapid functional enrichment analysis. Users
    can dynamically select genes from volcano or MA plots and instantly access
    enriched GO, KEGG, and Reactome pathway analyses, facilitating a deeper
    understanding of gene functions and biological pathways. The package
    supports customized plotting, feature information retrieval, and handles
    large datasets efficiently, making it an ideal tool for genomic
    researchers and bioinformaticians seeking to uncover the biological
    significance behind differential expression patterns.
vignette: >
  %\VignetteIndexEntry{Differential expression downstream analysis shiny app}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
options(width = 72)
library(tibble)
```

```{r setup}
library(dexdash)
```

# Purpose

# Installation

*dexdash* can be installed from *GitHub* by using:

```{r bioconductor_install, eval=FALSE}
if (!require("remotes", quietly = TRUE))
  install.packages("remotes")

remotes::install_github("bartongroup/dexdash")
```

# Quick example

```{r, eval=FALSE}
# Data examples available in the package
data(yeast_de, yeast_de, yeast_metadata, yeast_features)
# The slow bit: download functional term data
fterms <- download_functional_terms("yeast", feature_name = "gene_id")
# The fast bit: interactive app
run_app(yeast_de, yeast_de, yeast_metadata, yeast_features, fterms)
```

# Required input data

The input data for *dexdash* consists of 5 objects: `de`, `data`, `metadata`, `features` and `fterms`.

## Differential expression results

Argument name `de`. This is a data frame (or tibble) containing results of differential expression/abundance, per feature. It should have the following columns:

  - `id` (character) - feature identifier,
  - `log_fc` (numeric) - log-fold change,
  - `expr` (numeric) - mean expression or abundance,
  - `p_value` (numeric) - p-value from statistical tests,
  - `fdr` (numeric) - false discovery rate, p-value corrected for multiple tests,
  - `contrast` (factor) - name of the contrast; this data frame can contain results from multiple contrasts.
  
Here is an example of differential expression data, included in this package:

```{r}
data(yeast_de)
yeast_de
```

## Expression data

Argument name `data`. This is a data frame containing expression/abundance data, per sample, per feature. It should have the following columns:

 - `id` (character) - feature identifier, the same as in `de` data frame,
 - `sample` (character) - sample identifier,
 - `value` (numeric) - the expression, or abundance of the gene/protein for this sample.
 
Here is an example of expression data, included in this package:

```{r}
data(yeast_data)
yeast_data
```

## Design of the experiment
 
Argument name `metadata`. It contains the design of the experiment. It should have the following columns:
 
  - `sample` (character) - sample identifier, the same as in `data` data frame,
  - `group` (character/factor) - a grouping variable, for example condition or treatment,
  - `replicate` - optional replicate ID.

Here is an example of metadata, included in this package:

```{r}
data(yeast_metadata)
yeast_metadata
```

## Feature information

Argument name `features`. It contains alternative names for features (e.g. gene symbols) an descriptions. It should have the following columns:

 - `id` (character) - feature identifier, the same as in `de` and `data` frames,
 - `name` (character) - a human-readable name of the feature, typically a gene symbol,
 - `decription` (character) - a brief description of the feature.


Here is an example of feature information, included in this package:

```{r}
data(yeast_features)
yeast_features
```

## Functional term data

To perform fast interactive functional enrichment, the Shiny app requires data downloaded from online databases and prepared for fast enrichment. This can be achieved with the following function:

```{r, eval=FALSE}
fterms <- download_functional_terms(species = "yeast", feature_name = "gene_id")
```


The first argument in this function is a species name. Based on this name, the function will downlad the relevant data from the online databases. The GO, KEGG and Reactome databases use different species designation names. For example, designation for yeast is "sgd", "Saccharomyces cerevisiae" and "sce", for GO, Reactome and KEGG, respectively. In oder to interrogate these databases, the correct designations must be passed on. This package contains a small JSON file, which can be accessed:

```{r}
species_file <- system.file("extdata", "species.json", package = "dexdash")
```

Here is the content of the file:

```{r, results="asis", echo=FALSE}
cat("<pre>")
readLines(species_file) |> cat(sep = "\n")
cat("</pre>")
```

It contains designation information for a few species. If your species is not included, you need to create a JSON file in the same format, as the included file, and provide its location by argument `species_file` in the `download_functional_terms()` function. The species designations can be found using `fenr::fetch_go_species()`, `fenr::fetch_reactome_species()` and `fenr::fetch_kegg_species()`. These three functions return data frames,
where column `designation` contains the species designation required.

# Session info {.unnumbered}

```{r session_info, echo=FALSE}
sessionInfo()
```
